### DISCUSSION TITLE
It is possible knowing the size of a cell of any given grid?

### DISCUSSION THREAD

**Discussion by Cicorione** [speedata/publisher#539] [Q&A] [answered]
Hi @pgundlach 

I am trying to make a group unit the 85% of a cell of any given grid (3x3, 2x4, etc), is there a method to get the size of a cell?

I read there is this `sd:number-of-columns()`; can I use it to get something along the line:
```$unit-width={(sd:number-of-columns() div $ygrid) * 0.85}```

Thanks! 🙏

**Comment by pgundlach** [comment]
You can create a Group of size 1 x 1 (place a box with size 1 inside the group) and use `sd:group-width()`.

**Comment by pgundlach** [comment]
Another answer: if you are still inside this grid: you can just use `width="0.85"` for 85% of the grid cell width.

**Reply by Cicorione** [reply]
Can the children element of a group use proportional measurement as well? There is where I actually need to use proportional widths in order to have a layout more fluid, thanks!  🙏

**Reply by pgundlach** [reply]
You can always try yourself ;-)

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
    xmlns:sd="urn:speedata:2009/publisher/functions/en">

    <Record element="data">
        <Group name="a">
            <Grid width="5mm" height="5mm" />
            <Contents>
                <PlaceObject>
                    <Box height="1" width="1" />
                </PlaceObject>
            </Contents>
        </Group>
        <SetVariable variable="a-wd" select="sd:group-width('a','mm')"></SetVariable>
        <Group name="b">
            <Grid width="20mm" height="20mm" />
            <Contents>
                <PlaceObject>
                    <Box height="0.25" width="0.25" />
                </PlaceObject>
            </Contents>
        </Group>
        <SetVariable variable="b-wd" select="sd:group-width('b','mm')"></SetVariable>
        <PlaceObject>
            <Textblock>
                <Paragraph>
                    <Value select="$a-wd" />
                    <Value>, </Value>
                    <Value select="$b-wd" />
                </Paragraph>
            </Textblock>
        </PlaceObject>
    </Record>
</Layout>
```

Is that what you have in mind?

**Reply by Cicorione** [reply]
Actually I tried and I did something wrong then I asked for... 😳

**Comment by Cicorione** [comment]
Hi @pgundlach 

Using `width="0.85" for` inside the cell grid made the group smaller than 85%.

However using one of this two formulas with and without `{ }` curly brackets:

`sd:group-width`
```xml
<SetVariable variable="noc" select="sd:group-width('unit')" /> 
```
or

`sd:number-of-columns()`
```xml
<SetVariable variable="noc" select="sd:number-of-columns()" />
```
Targeting this:

```xml
<!-- The variable that control maxwidth -->
<SetVariable variable="mw" select="count(img)" />

<PlaceObject column="1" row="1">
	<Table width="{$noc * 0.85}" stretch="max" balance="yes">
		<Tr minheight="12">
			<ForAll select="img">
				<Td valign="middle" align="center">
					<Image file="{@file}" stretch="yes" maxheight="{12}"
                                               maxwidth="{($noc * 0.7) div ($mw * 1.2)}" />
				</Td>
			</ForAll>
		</Tr>
	</Table>
</PlaceObject>
```

Ended up with:
```sh
[...]
PlaceObject: Table at (1,1) wd/ht: 18/12 in "unit" (p. 5)
PlaceObject: Table at (1,14) wd/ht: 17/3 in "unit" (p. 5)
PlaceObject: Textblock at (1,17) wd/ht: 17/2 in "unit" (p. 5)
Error: [page 5] Could not convert dimension "\
   HDMI 19 pin male to HDMI 19 pin female right angle\
   24K Gold-plated\
   RoHS Compliant\
   \
"
/Git/publisher/src/lua/xpath.lua:632: attempt to perform arithmetic on a nil value (local 'second')
[...]
```

Sorry for all these questions, these days have been pretty catastrophic for my skills... 😭

**Reply by pgundlach** [reply]
I cannot reproduce the problem. 

```xml
<Layout xmlns="urn:speedata.de:2009/publisher/en"
  xmlns:sd="urn:speedata:2009/publisher/functions/en">

  <Record element="data">
    <SetVariable variable="noc" select="sd:number-of-columns()"></SetVariable>
    <!-- The variable that control maxwidth -->
    <SetVariable variable="mw" select="count(img)" />

    <PlaceObject column="1" row="1">
      <Table width="{$noc * 0.85}" stretch="max" balance="yes">
        <Tr minheight="12">
          <ForAll select="img">
            <Td valign="middle" align="center">
              <Image file="{@file}" stretch="yes" maxheight="12"
                maxwidth="{($noc * 0.7) div ($mw * 1.2)}" />
            </Td>
          </ForAll>
        </Tr>
      </Table>
    </PlaceObject>
  </Record>
</Layout>
```

and the following data file


```xml
<data>
    <img file="_samplea.pdf"/>
    <img file="_samplea.pdf"/>
    <img file="_samplea.pdf"/>
</data>
```

works fine here without problems. 

I think you need to strip down your layout to the aboslute minimum to reproduce the error and post it here again.

**Reply by Cicorione** [reply]
Uff... This is going to be harder than making it working... :thinking:

**Reply by pgundlach** [reply]
You can start by writing 

    <Message select="($noc * 0.7) div ($mw * 1.2)" />

after 

    <ForAll select="img">

for example and see if this shows something strange.

**Comment by Cicorione** [comment]
I got the a portion of the problem, it is related with the grid issue I sent you by email.

It fails when a group falls into a grid that it was supposed to use other rules. This issue that I have is shifting the grid to one page messing everything, I am pretty sure if I fix the grid shifting this will solve this issue.

**Reply by Cicorione** [reply, marked as answer]
I found issue I was doing math operation in this form: `$var * .8` rather than `$var * 0.8` 🤦‍♂️

However the math is totally wrong, it happened the exact contrary of what I thought, is taking the whole page... 😂
